astropy/units/distribution.py

```diff
@@ -0,0 +1,427 @@
+            if i >= maxtoshow:
+                break
+            if (not scalar_distr) and (not labelset):
+                if len(self.shape) == 2:
+                    idxstr = str(i)
+                else:
+                    idxstr = str(np.unravel_index(i, self.shape[1:]))
+                kwargs['label'] = 'Distr ' + idxstr
+            hists.append(visualization.hist(dat, **kwargs))
+        return hists
+
+
+class NormalDistribution(Distribution):
+    """
+    A Gaussian/normal Distribution
+
+    Parameters
+    ----------
+    center : `Quantity`
+        The center of this `NormalDistribution`
+    std : `Quantity` or `None`
+        The standard deviation/σ of this distribution. Shape must match and unit
+        must be compatible with ``center``, or be `None` (if ``var`` or ``ivar``
+        are set).
+    var : `Quantity` or `None`
+        The variance of this distribution. Shape must match and unit must be
+        compatible with ``center``, or be `None` (if ``std`` or ``ivar`` are set).
+    ivar : `Quantity` or `None`
+        The inverse variance of this distribution. Shape must match and unit
+        must be compatible with ``center``, or be `None` (if ``std`` or ``var``
+        are set).
+    n_samples : int
+        The number of Monte Carlo samples to use with this distribution
+
+    Remaining keywords are passed into the `Distribution` constructor
+    """
+    def __new__(cls, center, std=None, var=None, ivar=None, n_samples=1000, **kwargs):
+        center = np.asanyarray(center)
+        if var is not None:
+            if std is None:
+                std = np.asanyarray(var)**0.5
+            else:
+                raise ValueError('NormalDistribution cannot take both std and var')
+        if ivar is not None:
+            if std is None:
+                std = np.asanyarray(ivar)**-0.5
+            else:
+                raise ValueError('NormalDistribution cannot take both ivar and '
+                                 'and std or var')
+        if std is None:
+            raise ValueError('NormalDistribution requires one of std, var, or ivar')
+        else:
+            std = np.asanyarray(std)
+
+        randshape = np.broadcast(std, center).shape + (n_samples,)
+        distr = center[..., np.newaxis] + np.random.randn(*randshape) * std[..., np.newaxis]
+        self = super(NormalDistribution, cls).__new__(cls, distr, **kwargs)
+        self.distr_std = std
+        return self
+
+
+class PoissonDistribution(Distribution):
+    """
+    A Poisson Distribution
+
+    Parameters
+    ----------
+    poissonval : `Quantity`
+        The center value of this `PoissonDistribution` (i.e., λ).
+    n_samples : int
+        The number of Monte Carlo samples to use with this distribution
+
+    Remaining keywords are passed into the `Distribution` constructor
+    """
+    def __new__(cls, poissonval, n_samples=1000, **kwargs):
+        if hasattr(poissonval, 'unit') and 'unit' not in kwargs:
```

Comment:
Not sure I like allowing any unit except for counts (the only one that makes sense). For now, I would not even allow any unit.